File: src/textbook-loader/citations/resolvers/research-db.ts
/**
 * The research-database resolver — the local, free, try-first accelerator.
 *
 * The homelab corpus (4,7+K articles of frontier-AI and AI-safety source
 * material) answers by exact URL with CSL-grade metadata and no third-party
 * etiquette to observe, so it is tried before any networked resolver. It is
 * still only an accelerator, never a dependency: coverage is a minority of the
 * corpus (measured 2026-09-21, 78 of 987 cited URLs), and `task:0021` D4
 * requires the bibliography to build identically with the service down —
 * hence AC-6 below.
 *
 * Three traps this file exists to handle, all discovered against the live
 * service:
 *
 * 1. **Admission scope.** Records default to `admitted` and many relevant ones
 *    are `unreviewed`, so every request widens with `include_unreviewed=true`.
 *    Without it, resolution quietly misses records the corpus holds.
 * 2. **URL spelling.** The corpus stores scraped URLs verbatim (modulo scheme/
 *    host case), and matching is exact string equality (`records_for_url`).
 *    ~39% of stored URLs carry `www.` while our canonical form drops it, which
 *    silently loses 27 of the 987 cited URLs — so a miss retries once with
 *    `www.` inserted. Trailing-slash differences are handled server-side;
 *    nothing else is aliased, and inventing more variants than that would be
 *    guessing at corpus state.
 * 3. **Authorship shape.** `authors[]` entries are `{kind, label, resolved}`
 *    bylines, not family/given — and a *resolved organisation* author has no
 *    `label` at all. Labels map to CSL `literal` names and nulls are dropped;
 *    guessing a split here would produce confidently wrong output in every
 *    rendered style (see `task:0027`).
 *
 * AC-6 is structural: an unreachable service, a non-200 response, or an
 * unusable body all return null — never a throw, never a hang — so a resolver
 * that accelerates 8% of sources cannot become a single point of failure for
 * the other 92%.
 */
import type { CslDate, CslName, CslType } from '../store.js';
import { literalName } from '../store.js';
import type { ResolveResult, Resolver, ResolverContext } from './types.js';

/** Same variable the `research` CLI honours, so one setting controls both. */
function apiBase(): string {
  return process.env.RESEARCH_API || 'http://localhost:8551';
}

/** Generous for a LAN service, short enough that a stalled one never wedges a run. */
const TIMEOUT_MS = 5_000;

/** The shape of `GET /api/records/citation` on success. Only the fields we map. */
type CitationPayload = {
  success?: boolean;
  title?: string | null;
  publication_date?: string | null;
  abstract?: string | null;
  source_type?: string | null;
  authors?: Array<{ label?: string | null } | null>;
};

/**
 * `source_type` is a provenance-first label, not a CSL type. The corpus's
 * types are organisations (anthropic, deepmind, epoch…) plus a few formats.
 * Only the mappings that are actually distinctive are made here; everything
 * else renders as a webpage, which is what an org landing page is.
 */
const SOURCE_TYPE_TO_CSL: Record<string, CslType> = {
  arxiv: 'article',
  substack: 'post-weblog',
  alignmentforum: 'post-weblog',
  lesswrong: 'post-weblog',
  youtube: 'motion_picture',
};

/** `2024-08-15T17:23:10Z` (or a bare date) → CSL date-parts, or null. */
function parseDate(iso: string | null | undefined): CslDate | null {
  const m = /^(\d{4})(?:-(\d{2}))?(?:-(\d{2}))?/.exec(iso ?? '');
  if (!m) return null;
  const parts = [Number(m[1]), m[2] ? Number(m[2]) : undefined, m[3] ? Number(m[3]) : undefined];
  return { 'date-parts': [parts.filter((p) => p !== undefined)] };
}

/** Captured bylines → CSL literal names. A resolved-organisation author (null label) is not a name. */
function mapAuthors(authors: CitationPayload['authors']): CslName[] {
  return (authors ?? [])
    .filter((a): a is { label: string } => typeof a?.label === 'string' && a.label.trim() !== '')
    .map((a) => literalName(a.label));
}

/**
 * One request, one verdict. Returns null for every "tried, found nothing"
 * outcome: network error, non-200, non-JSON body, `success` false, or a record
 * too empty to improve the entry. It does not throw, which is what makes the
 * two-variant loop below safe.
 */
async function attempt(refUrl: string, ctx: ResolverContext): Promise<ResolveResult | null> {
  const url =
    `${apiBase()}/api/records/citation?ref=${encodeURIComponent(refUrl)}` +
    '&include_unreviewed=true';
  let payload: CitationPayload;
  try {
    const response = await ctx.fetch(url, {
      headers: { 'User-Agent': ctx.userAgent },
      signal: AbortSignal.timeout(TIMEOUT_MS),
    });
    if (!response.ok) return null;
    payload = (await response.json()) as CitationPayload;
  } catch {
    // Unreachable, timed out, or garbage body. This is the AC-6 path: null,
    // not an exception — rule 2 of the resolver contract.
    return null;
  }
  if (!payload.success) return null;

  const fields: ResolveResult['fields'] = {};
  if (typeof payload.title === 'string' && payload.title.trim()) {
    fields.title = payload.title;
  }
  const authors = mapAuthors(payload.authors);
  if (authors.length > 0) fields.author = authors;
  const issued = parseDate(payload.publication_date);
  if (issued) fields.issued = issued;
  if (typeof payload.abstract === 'string' && payload.abstract.trim()) {
    fields.abstract = payload.abstract;
  }
  if (payload.source_type && SOURCE_TYPE_TO_CSL[payload.source_type]) {
    fields.type = SOURCE_TYPE_TO_CSL[payload.source_type];
  }
  if (Object.keys(fields).length === 0) return null;

  const missing: string[] = [];
  if (authors.length === 0) missing.push('no author labels');
  if (!issued) missing.push('no publication date');
  // Deliberately factual, in the store's note voice: what is absent, not a
  // warranty about what is present. The corpus leaves attribution
  // `legacy_unknown` until reviewed; the labels it returns are the captured
  // byline, which is more than the anchor text knew but not more than that.
  return {
    fields,
    source: 'research-db',
    note:
      missing.length > 0
        ? `Research-database record found, but ${missing.join(' and ')}.`
        : undefined,
  };
}

/** Canonical URLs are https and have `www.` stripped; claims() covers all of them. */
function withoutWww(url: string): boolean {
  try {
    return !new URL(url).hostname.startsWith('www.');
  } catch {
    return false;
  }
}

function withWww(url: string): string | null {
  try {
    const u = new URL(url);
    u.hostname = `www.${u.hostname}`;
    return u.toString();
  } catch {
    return null;
  }
}

export const researchDbResolver: Resolver = {
  name: 'research-db',

  // Broadly true by design: the corpus spans many domains, so almost any URL
  // is worth one local request before the networked resolvers are paid for.
  claims(canonicalUrl: string): boolean {
    return /^https?:\/\//.test(canonicalUrl);
  },

  async resolve(canonicalUrl: string, ctx: ResolverContext): Promise<ResolveResult | null> {
    const result = await attempt(canonicalUrl, ctx);
    if (result) return result;

    if (withoutWww(canonicalUrl)) {
      const www = withWww(canonicalUrl);
      if (www) return attempt(www, ctx);
    }
    return null;
  },
};

File: python/atlas_citations/resolvers/research_db.py
"""The research-database resolver — the local, free, try-first accelerator.

The homelab corpus answers by exact URL with CSL-grade metadata and no
third-party etiquette to observe, so it is tried before any networked resolver.
It is still only an accelerator, never a dependency: coverage is a minority of
the corpus (measured 2026-09-21 at 7.9% of cited URLs), and ``task:0021`` D4
requires the bibliography to build identically with the service down.

Three traps this file exists to handle, all discovered against the live service:

1. **Admission scope.** Records default to ``admitted`` and many relevant ones
   are ``unreviewed``, so every request widens with ``include_unreviewed=true``.
   Without it, resolution quietly misses records the corpus holds.
2. **URL spelling.** The corpus stores scraped URLs verbatim (modulo scheme/host
   case) and matching is exact string equality. Roughly 39% of stored URLs carry
   ``www.`` while our canonical form drops it, which silently loses cited URLs —
   so a miss retries once with ``www.`` inserted. Trailing-slash differences are
   handled server-side; nothing else is aliased, and inventing more variants
   would be guessing at corpus state.
3. **Authorship shape.** ``authors[]`` entries are ``{kind, label, resolved}``
   bylines, not family/given — and a *resolved organisation* author has no
   ``label`` at all. Labels map to CSL ``literal`` names and nulls are dropped;
   guessing a split here would produce confidently wrong output in every
   rendered style.

AC-6 is structural: an unreachable service, a non-200 response, or an unusable
body all return ``None`` — never a raise, never a hang — so a resolver that
accelerates a minority of sources cannot become a single point of failure for the
rest.

Ported from ``research-db.ts`` under ``task:0029``.
"""

from __future__ import annotations

import os
import re
from typing import Any
from urllib.parse import quote

from ..store import literal_name, with_www
from ._http import get_json
from .base import ResolverContext, ResolveResult

#: Generous for a LAN service, short enough that a stalled one never wedges a run.
TIMEOUT_S = 5.0

#: ``source_type`` is a provenance-first label, not a CSL type. The corpus's types
#: are organisations (anthropic, deepmind, epoch…) plus a few formats. Only the
#: mappings that are actually distinctive are made here; everything else renders
#: as a webpage, which is what an org landing page is.
SOURCE_TYPE_TO_CSL = {
    "arxiv": "article",
    "substack": "post-weblog",
    "alignmentforum": "post-weblog",
    "lesswrong": "post-weblog",
    "youtube": "motion_picture",
}

_DATE = re.compile(r"^(\d{4})(?:-(\d{2}))?(?:-(\d{2}))?")


def api_base() -> str:
    """Same variable the ``research`` CLI honours, so one setting controls both."""
    return os.environ.get("RESEARCH_API") or "http://localhost:8551"


def _parse_date(raw: Any) -> dict[str, Any] | None:
    """``2024-08-15T17:23:10Z`` (or a bare date) → CSL date-parts, or ``None``."""
    if not isinstance(raw, str):
        return None
    m = _DATE.match(raw)
    if not m:
        return None
    return {"date-parts": [[int(g) for g in m.groups() if g is not None]]}


def _map_authors(authors: Any) -> list[dict[str, str]]:
    """Captured bylines → CSL literal names.

    A resolved-organisation author (null label) is not a name.
    """
    if not isinstance(authors, list):
        return []
    out = []
    for author in authors:
        if isinstance(author, dict):
            label = author.get("label")
            if isinstance(label, str) and label.strip():
                out.append(literal_name(label))
    return out


def _attempt(ref_url: str, ctx: ResolverContext) -> ResolveResult | None:
    """One request, one verdict.

    Returns ``None`` for every "tried, found nothing" outcome: network error,
    non-200, non-JSON body, ``success`` false, or a record too empty to improve
    the entry. It does not raise, which is what makes the two-variant retry safe.

    Deliberately flattens ``task:0032``'s unreachable signal back to ``None``:
    this corpus is a local accelerator, and ``task:0027`` AC-6 requires the
    bibliography to come out identical when it is down. An outage here must be
    invisible, not a verdict — which is also why the resolver is ``selective =
    False`` and so cannot block the fallback chain.
    """
    url = f"{api_base()}/api/records/citation?ref={quote(ref_url, safe='')}&include_unreviewed=true"
    payload = get_json(ctx, url, timeout=TIMEOUT_S)
    if not isinstance(payload, dict) or not payload.get("success"):
        return None

    fields: dict[str, Any] = {}
    title = payload.get("title")
    if isinstance(title, str) and title.strip():
        fields["title"] = title
    authors = _map_authors(payload.get("authors"))
    if authors:
        fields["author"] = authors
    issued = _parse_date(payload.get("publication_date"))
    if issued:
        fields["issued"] = issued
    abstract = payload.get("abstract")
    if isinstance(abstract, str) and abstract.strip():
        fields["abstract"] = abstract
    source_type = payload.get("source_type")
    if source_type in SOURCE_TYPE_TO_CSL:
        fields["type"] = SOURCE_TYPE_TO_CSL[source_type]

    if not fields:
        return None

    missing = []
    if not authors:
        missing.append("no author labels")
    if not issued:
        missing.append("no publication date")
    # Deliberately factual, in the store's note voice: what is absent, not a
    # warranty about what is present. The corpus leaves attribution
    # `legacy_unknown` until reviewed; the labels it returns are the captured
    # byline, which is more than the anchor text knew but not more than that.
    note = f"Research-database record found, but {' and '.join(missing)}." if missing else None
    return ResolveResult(fields=fields, source="research-db", note=note)


class ResearchDbResolver:
    name = "research-db"
    #: Claims every HTTP URL — a free local lookup, not a signal of coverage.
    selective = False

    def claims(self, canonical_url: str) -> bool:
        # Broadly true by design: the corpus spans many domains, so almost any
        # URL is worth one local request before the networked resolvers are
        # paid for.
        return canonical_url.startswith(("http://", "https://"))

    def resolve(self, canonical_url: str, ctx: ResolverContext) -> ResolveResult | None:
        result = _attempt(canonical_url, ctx)
        if result:
            return result
        www = with_www(canonical_url)
        return _attempt(www, ctx) if www else None


research_db_resolver = ResearchDbResolver()
